### PR TITLE
kola: Add argument to enable channel based testing

### PR DESCRIPTION
**Pull request by sayanchowdhury**
Add argument to enable channel based testing

This PR adds the argument `--channel` to enable channel based testing, which
means you can populate the Channel property in kola tests and kola would decide to
include/exclude the test based on which channel it is running.

This PR also output the channels when the list subcommand is run both with or without `--json`

# How to use

`./bin/kola run [--channel <channel>]`
`./bin/kola list [--channel <channel>]`

# Testing done

Both the outputs are snipped.

```
$ ./bin/kola list
Test Name                                       Platforms       Architectures   Distributions   Channels

cl.basic                                        [all]                                                   [all]   [cl]            [all]
cl.cloudinit.basic                              [aws azure do esx gce openstack packet qemu]            [all]   [cl]            [all]
cl.cloudinit.script                             [aws azure do esx gce openstack packet qemu]            [all]   [cl]            [all]
cl.disk.raid.data                               [all]                                                   [all]   [cl]            [all]
cl.disk.raid.root                               [qemu]                                                  [all]   [cl]            [all]
cl.etcd-member.discovery                        [all]                                                   [all]   [cl]            [all]
cl.etcd-member.etcdctlv3                        [all]                                                   [all]   [cl]            [all]
cl.etcd-member.v2-backup-restore                [aws azure do gce openstack packet qemu qemu-unpriv]    [all]   [cl]            [all]
cl.filesystem                                   [all]                                                   [all]   [cl]            [all]
cl.flannel.udp                                  [all]                                                   [all]   [cl]            [all]
cl.flannel.vxlan                                [all]                                                   [all]   [cl]            [all]
cl.ignition.misc.empty                          [aws azure do gce openstack packet qemu-unpriv]         [all]   [cl]            [all]
cl.ignition.v1.btrfsroot                        [all]                                                   [all]   [cl]            [all]
cl.ignition.v1.ext4root                         [all]                                                   [all]   [cl]            [all]
cl.ignition.v1.groups                           [all]                                                   [all]   [cl]            [all]
cl.ignition.v1.noop                             [aws azure do gce packet qemu-unpriv]                   [all]   [cl]            [all]
cl.ignition.v1.once                             [all]                                                   [all]   [cl]            [all]
cl.ignition.v1.sethostname                      [aws do esx gce openstack packet qemu qemu-unpriv]      [all]   [cl]            [all]
cl.ignition.v1.ssh.key                          [aws azure do esx gce openstack packet qemu-unpriv]     [all]   [cl]            [all]
cl.ignition.v1.users                            [all]                                                   [all]   [cl]            [all]
cl.ignition.v1.xfsroot                          [all]                                                   [all]   [cl]            [all]
cl.ignition.v2.btrfsroot                        [all]                                                   [all]   [cl]            [all]
cl.ignition.v2.ext4root                         [all]                                                   [all]   [cl]            [all]
cl.ignition.v2.noop                             [aws azure do gce packet qemu-unpriv]                   [all]   [cl]            [all]
cl.ignition.v2.users                            [all]                                                   [all]   [cl]            [all]
cl.ignition.v2.xfsroot                          [all]                                                   [all]   [cl]            [all]
cl.ignition.v2_1.ext4checkexisting              [all]                                                   [all]   [cl]            [all]
cl.ignition.v2_1.swap                           [all]                                                   [all]   [cl]            [all]
cl.ignition.v2_1.vfat                           [all]                                                   [all]   [cl]            [all]
cl.install.cloudinit                            [aws do esx gce openstack packet qemu qemu-unpriv]      [all]   [cl]            [all]
cl.internet                                     [all]                                                   [all]   [cl]            [all]
cl.locksmith.cluster                            [all]                                                   [all]   [cl]            [all]
cl.metadata.aws                                 [aws]                                                   [all]   [cl]            [all]
cl.metadata.azure                               [azure]                                                 [all]   [cl]            [all]
cl.metadata.packet                              [packet]                                                [all]   [cl]            [all]
cl.network.initramfs.second-boot                [aws azure esx gce openstack packet qemu qemu-unpriv]   [all]   [cl]            [all]
cl.network.listeners                            [aws azure do esx gce openstack packet qemu]            [all]   [cl]            [all]
cl.network.listeners.legacy                     [aws azure do esx gce openstack packet qemu]            [all]   [cl]            [all]
cl.omaha.ping                                   [qemu]                                                  [all]   [cl]            [all]
cl.rkt.etcd3                                    [all]                                                   [all]   [cl]            [all]
cl.toolbox.dnf-install                          [all]                                                   [all]   [cl]            [all]
cl.update.badverity                             [all]                                                   [all]   [cl]            [all]
cl.update.grubnop                               [aws azure do esx gce openstack packet qemu]            [amd64] [cl]            [all]
cl.update.payload                               [all]                                                   [all]   [cl]            [all]
cl.update.reboot                                [all]                                                   [all]   [cl]            [all]
cl.users.shells                                 [aws azure do esx openstack packet qemu qemu-unpriv]    [all]   [cl]            [all]
cl.verity                                       [all]                                                   [all]   [cl]            [all]
coreos.auth.verify                              [all]                                                   [all]   [cl fcos rhcos] [all]
coreos.ignition.groups                          [all]                                                   [all]   [cl fcos rhcos] [all]
coreos.ignition.once                            [all]                                                   [all]   [cl fcos rhcos] [all]
```

```
$ ./bin/kola list --json
[
        {
                "Name": "cl.basic",
                "Platforms": [
                        "aws",
                        "azure",
                        "do",
                        "esx",
                        "gce",
                        "openstack",
                        "packet",
                        "qemu",
                        "qemu-unpriv"
                ],
                "Architectures": [
                        "amd64"
                ],
                "Distros": [
                        "cl"
                ],
                "Channels": [
                        "alpha",
                        "beta",
                        "stable",
                        "edge"
                ]
        },
        {
                "Name": "cl.cloudinit.basic",
                "Platforms": [
                        "aws",
                        "azure",
                        "do",
                        "esx",
                        "gce",
                        "openstack",
                        "packet",
                        "qemu"
                ],
                "Architectures": [
                        "amd64"
                ],
                "Distros": [
                        "cl"
                ],
                "Channels": [
                        "alpha",
                        "beta",
                        "stable",
                        "edge"
                ]
        },
```

### CODE REVIEW

File: kola/register/register.go [hunk 1]
```diff
@@ -56,6 +56,8 @@ type Test struct {
 	ExcludePlatforms []string // blacklist of platforms to ignore -- defaults to none
 	Distros          []string // whitelist of distributions to run test against -- defaults to all
 	ExcludeDistros   []string // blacklist of distributions to ignore -- defaults to none
+	Channels         []string // whitelist of channels to run test against -- defaults to all
+	ExcludeChannels  []string // blacklist of channels to ignore -- defaults to all
```
Comment (marga-kinvolk) [inline kola/register/register.go:60]: Shouldn't this one default to none?

File: cmd/kola/options.go [hunk 2]
```diff
@@ -64,6 +66,7 @@ func init() {
 	sv(&outputDir, "output-dir", "", "Temporary output directory for test data and logs")
 	sv(&kola.TorcxManifestFile, "torcx-manifest", "", "Path to a torcx manifest that should be made available to tests")
 	root.PersistentFlags().StringVarP(&kolaPlatform, "platform", "p", "qemu", "VM platform: "+strings.Join(kolaPlatforms, ", "))
+	root.PersistentFlags().StringVarP(&kolaChannel, "channel", "c", "edge", "Channel: "+strings.Join(kolaChannels, ", "))
```
Comment (pothos) [inline cmd/kola/options.go:69]: I'm not sure if having `edge` as default makes sense. I think we would set it explicitly anyway and for users `stable` would be more common, or?

Comment (marga-kinvolk) [reply cmd/kola/options.go:69]: Are stable users running kola tests?

File: kola/register/register.go (hunk 1 above)
Comment (sayanchowdhury) [reply kola/register/register.go:60]: Oh! yes.

File: cmd/kola/options.go (hunk 2 above)
Comment (dongsupark) [reply cmd/kola/options.go:69]: +1 on making `stable` by default.

**Comment by dongsupark** [comment]
Nit: the first commit says `Fixes #42`, but it should actually point to https://github.com/flatcar-linux/Flatcar/issues/42.

File: cmd/kola/options.go (hunk 2 above)
Comment (sayanchowdhury) [reply cmd/kola/options.go]: Agreed @pothos, the channels have to be explicitly mentioned while  we are  our running the tests. I've updated the default to `stable`.

**Comment by sayanchowdhury** [comment]
The commit message is updated @dongsupark

**Pull request merged**